features/lorawan/lorastack/phy/LoRaPHY.cpp

```diff
@@ -122,316 +100,1315 @@ void LoRaPHY::handle_send(uint8_t *buf, uint8_t size)
+        *time = next_tx_delay;
+        return true;
+    }
+
+    // Datarate not supported by any channel, restore defaults
+    copy_channel_mask(phy_params.channels.mask,
+                      phy_params.channels.default_mask,
+                      phy_params.channels.mask_size);
+    *time = 0;
+    return false;
+}
+
+lorawan_status_t LoRaPHY::add_channel(channel_params_t* new_channel, uint8_t id)
+{
+    bool dr_invalid = false;
+    bool freq_invalid = false;
+
+    if (!phy_params.custom_channelplans_supported
+            || id >= phy_params.max_channel_cnt) {
+
+        return LORAWAN_STATUS_PARAMETER_INVALID;
+    }
+
+    // Validate the datarate range
+    if (val_in_range(new_channel->dr_range.fields.min,
+                     phy_params.min_tx_datarate,
+                     phy_params.max_tx_datarate) == 0) {
+        dr_invalid = true;
+    }
+
+    if (val_in_range(new_channel->dr_range.fields.max, phy_params.min_tx_datarate,
+                     phy_params.max_tx_datarate) == 0) {
+        dr_invalid = true;
+    }
+
+    if (new_channel->dr_range.fields.min > new_channel->dr_range.fields.max) {
+        dr_invalid = true;
+    }
+
+    // Default channels don't accept all values
+    if (id < phy_params.default_channel_cnt) {
+        // Validate the datarate range for min: must be DR_0
+        if (new_channel->dr_range.fields.min > phy_params.min_tx_datarate) {
+            dr_invalid = true;
+        }
+
+        // Validate the datarate range for max: must be DR_5 <= Max <= TX_MAX_DATARATE
+        if (val_in_range(new_channel->dr_range.fields.max,
+                         phy_params.default_max_datarate,
+                         phy_params.max_tx_datarate) == 0) {
+            dr_invalid = true;
+        }
+
+        // We are not allowed to change the frequency
+        if (new_channel->frequency != phy_params.channels.channel_list[id].frequency) {
+            freq_invalid = true;
+        }
+    }
+
+    // Check frequency
+    if (!freq_invalid) {
+        if (verify_frequency(new_channel->frequency) == false) {
+            freq_invalid = true;
+        }
+    }
+
+    // Check status
+    if (dr_invalid && freq_invalid) {
+        return LORAWAN_STATUS_FREQ_AND_DR_INVALID;
+    }
+
+    if (dr_invalid) {
+        return LORAWAN_STATUS_DATARATE_INVALID;
+    }
+
+    if (freq_invalid) {
+        return LORAWAN_STATUS_FREQUENCY_INVALID;
+    }
+
+    memcpy(&(phy_params.channels.channel_list[id]), new_channel, sizeof(channel_params_t));
+
+    phy_params.channels.channel_list[id].band = new_channel->band;
+
+    // if there are multiple channel masks, i.e., there are more than 16 channels
+    // defined by the PHY layer, we search the channel index in all masks and
+    // set the appropriate bit in the appropriate mask
+    for (uint8_t i = 0; i < phy_params.max_channel_cnt; i ++) {
+        if (i == id) {
+            MASK_BIT_SET(phy_params.channels.mask[i/16], (i % 16));
+        }
+    }
```

Comment:
This is weird. You could just do:

    MASK_BIT_SET(phy_params.channels.mask[id/16], (id % 16));

Or, more conventionally:

    MASK_BIT_SET(phy_params.channels.mask, id);

If MASK_BIT_SET was a general bit array processing macro/function, which is what I was after. At the minute, the macro is a little pointless.

Do you have an objection to this form? 

    #define MASK_BIT_TEST(mask, bit)    ((mask)[(bit) / 16] & (1U << [(bit) % 16]))

Reply:
'if (i == id) {'  this bit is there mistake. It shouldn't be there. The form that you are suggesting makes the macro workable with 16 bit blocks only. Yes, we don't have any other blocks to handle at the moment but I just wanted to keep my options open.

Reply:
> The form that you are suggesting makes the macro workable with 16 bit blocks only.

That's the point. Every piece of code manipulating the mask currently has to know the channel mask is using 16-bit blocks (and not make a mistake). The point of the macros was to abstract that implementation detail into one central place - the macro definitions.

If you had a different mask format, you'd use a different helper set to manipulate it. (eg the big-endian bytes version in common_functions.h). You could call yours `CHANNEL_MASK_BIT_xxx` to make clear that's what they're for.

Reply:
Above macro doesn't work. Invalid type for array subscript. I could make an inline function to do this.

Reply:
How about this ?

```
     inline uint8_t mask_set_bit(uint16_t *mask, uint8_t bit) {
         return mask[bit/16] & (1U << (bit % 16));
     }
```

Reply:
Yes, functions would be better than macros.

For small almost-macro functions like this I suggest `static inline` to avoid various annoyances with C/C++ incompatibility. But if the header file is C++-only, don't worry.

The `bit` argument probably should be an int or unsigned int - no particular need to limit it to 256 channels? Narrower-than-int arguments and variables tend to generate extra code on the ARM. Only generally worthwhile if they're part of a structure.

Assuming that's supposed to be the bit test, it should return `bool`, and take a `const` mask pointer.

Certainly that `uint8_t` return type won't work - if you were testing bit 9 in a word you'd get `0x0200` from `&` then narrow it to `0x00`.

A `bool` return would convert neatly automatically - it will implicitly convert any non-zero value to `true`/`1`.